### DISCUSSION TITLE
How do edge_index multiply matrice？

### DISCUSSION THREAD

**Discussion by mumu029** [pyg-team/pytorch_geometric#5164] [Q&A] [answered]
```python
def forward(self,x,edge_index):
        z = x
        if last_layer == False:
            s = self.pool(x,edge_index)
        x_new = s.T @ z
        a_new = s.T @ edge_index @ s
        return x_new,a_new
```
I wanted to implement the DiffPool method, but I found that in torch_feometric, edge_index is a vector of size [2,num_edge], and S is a vector of size [num_node,new_num_node]. They can't multiply. What should I do?

**Comment by rusty1s** [comment, marked as answer]
`edge_index` needs to be transformed first into a dense adjacency matrix via `utils.to_dense_adj`. Alternatively, you can convert `edge_index` into a `SparseTensor` and do
```python
from torch_sparse import SparseTensor
adj = SparseTensor.from_edge_index(edge_index)
a_new = s.T @ (adj @ s)
```

**Reply by mumu029** [reply]
Thank you very much for your answer, it was very helpful. However, I still have a question: why does dense_diff_pool() accept an adjacency matrix instead of an edge_index? This causes edge_index to be converted to an adjacency matrix before using this function, and then changed back again after using this function.

**Reply by rusty1s** [reply]
`dense_diff_pool` is a dense operator/it is only really defined on dense adjacency matrices. Importantly, the dense representation is never converted back since we no longer can utilize the sparsity in the underlying output representation (it is more of a probability matrix now whether a link exists). As such, if you want to stack dense diff pool layers, you need to operate on dense adjacency matrices anyway.

**Reply by mumu029** [reply]
Well, thank you very much for your answer.
